File: daemon/telnet.c
/*
 * SerialTCP -- a virtual serial device and modem daemon for AmigaOS.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, see <https://www.gnu.org/licenses/>.
 */

/*
 * telnet.c -- RFC 854 option negotiation and data transparency.
 *
 * Two jobs:
 *
 *  1. Negotiate BINARY in both directions.  This is not optional for a BBS:
 *     without it Zmodem transfers corrupt on any byte that happens to be 0xFF,
 *     and high-bit ANSI art gets mangled.
 *
 *  2. Keep the data path transparent.  0xFF on the wire means IAC, so a real
 *     0xFF in the data stream has to be doubled, and in NVT (non-binary) mode a
 *     bare CR must be followed by NUL.
 *
 * Negotiation refusals are answered exactly once per request and we never
 * answer a WONT/DONT that does not change our state, which is what keeps the
 * classic negotiation ping-pong loop from happening.
 */

#include <exec/types.h>
#include <string.h>

#include "daemon.h"

/* Options we are willing to turn on.  Everything else is refused. */
static BOOL we_support_local(struct Telnet *t, UBYTE opt)
{
    switch (opt)
    {
    case TELOPT_BINARY: return TRUE;
    case TELOPT_SGA:    return TRUE;
    case TELOPT_ECHO:   return t->t_Server;   /* only a server offers to echo */
    default:            return FALSE;
    }
}

static BOOL we_support_remote(struct Telnet *t, UBYTE opt)
{
    switch (opt)
    {
    case TELOPT_BINARY: return TRUE;
    case TELOPT_SGA:    return TRUE;
    case TELOPT_ECHO:   return !t->t_Server;  /* a client lets the server echo */
    default:            return FALSE;
    }
}

/* Stage a three-byte negotiation command for the caller to send. */
static void send_cmd(struct Telnet *t, UBYTE verb, UBYTE opt)
{
    if (t->t_RespLen + 3 > SUBNEG_SIZE)
        return;                         /* flooded: decline to answer */

    t->t_RespBuf[t->t_RespLen++] = TEL_IAC;
    t->t_RespBuf[t->t_RespLen++] = verb;
    t->t_RespBuf[t->t_RespLen++] = opt;
}

ULONG telnet_take_responses(struct Telnet *t, UBYTE *out, ULONG outmax)
{
    ULONG n = t->t_RespLen;

    if (n > outmax)
        n = outmax;

    if (n)
        memcpy(out, t->t_RespBuf, n);

    /* Anything that would not fit is dropped rather than left half-sent: a
     * partial IAC sequence would desynchronise the peer. */
    t->t_RespLen = 0;
    return n;
}

void telnet_reset(struct Telnet *t, BOOL enabled, BOOL server)
{
    memset(t, 0, sizeof(*t));
    t->t_Enabled = enabled;
    t->t_Server  = server;
    t->t_State   = TS_DATA;
}

/* Opening offer, sent as soon as a connection is established. */
void telnet_start_t(struct Telnet *t)
{
    if (!t->t_Enabled || t->t_SentInitial)
        return;

    t->t_SentInitial = TRUE;

    if (t->t_Server)
    {
        send_cmd(t, TEL_WILL, TELOPT_ECHO);
        send_cmd(t, TEL_WILL, TELOPT_SGA);
    }
    else
    {
        send_cmd(t, TEL_DO, TELOPT_SGA);
    }

    /* Ask for 8-bit clean in both directions. */
    send_cmd(t, TEL_WILL, TELOPT_BINARY);
    send_cmd(t, TEL_DO,   TELOPT_BINARY);
}

static void handle_will(struct Telnet *t, UBYTE opt)
{
    if (we_support_remote(t, opt))
    {
        if (opt == TELOPT_BINARY)
        {
            if (t->t_BinaryRemote)
                return;                 /* already on: stay quiet */
            t->t_BinaryRemote = TRUE;
        }
        send_cmd(t, TEL_DO, opt);
    }
    else
    {
        send_cmd(t, TEL_DONT, opt);
    }
}

static void handle_wont(struct Telnet *t, UBYTE opt)
{
    if (opt == TELOPT_BINARY)
    {
        if (!t->t_BinaryRemote)
            return;                     /* already off: no state change */
        t->t_BinaryRemote = FALSE;
    }

    send_cmd(t, TEL_DONT, opt);
}

static void handle_do(struct Telnet *t, UBYTE opt)
{
    if (we_support_local(t, opt))
    {
        if (opt == TELOPT_BINARY)
        {
            if (t->t_BinaryLocal)
                return;
            t->t_BinaryLocal = TRUE;
        }
        send_cmd(t, TEL_WILL, opt);
    }
    else
    {
        send_cmd(t, TEL_WONT, opt);
    }
}

static void handle_dont(struct Telnet *t, UBYTE opt)
{
    if (opt == TELOPT_BINARY)
    {
        if (!t->t_BinaryLocal)
            return;
        t->t_BinaryLocal = FALSE;
    }

    send_cmd(t, TEL_WONT, opt);
}

/*
 * Wire bytes in, application bytes out.
 *
 * Negotiation replies are staged in the Telnet state for the caller to drain,
 * so they never pass through the encoder and never get IAC-doubled.
 *
 * Returns the number of application bytes written to `out`.  Never writes more
 * than inlen bytes, so `outmax >= inlen` is always sufficient.
 */
ULONG telnet_decode_t(struct Telnet *t, const UBYTE *in, ULONG inlen, UBYTE *out, ULONG outmax)
{
    ULONG          o = 0;
    ULONG          i;
    UBYTE          c;

    if (!t->t_Enabled)
    {
        ULONG len = (inlen < outmax) ? inlen : outmax;
        memcpy(out, in, len);
        return len;
    }

    for (i = 0; i < inlen && o < outmax; i++)
    {
        c = in[i];

        switch (t->t_State)
        {
        case TS_DATA:
            if (c == TEL_IAC)
            {
                t->t_State = TS_IAC;
            }
            else if (!t->t_BinaryRemote && c == 0 && t->t_OutLastCR)
            {
                /* NVT CR NUL means a bare CR.  Swallow the NUL. */
                t->t_OutLastCR = FALSE;
            }
            else
            {
                t->t_OutLastCR = (!t->t_BinaryRemote && c == '\r');
                out[o++] = c;
            }
            break;

        case TS_IAC:
            switch (c)
            {
            case TEL_IAC:                       /* escaped 0xFF */
                out[o++]       = TEL_IAC;
                t->t_OutLastCR = FALSE;
                t->t_State     = TS_DATA;
                break;
            case TEL_WILL: t->t_State = TS_WILL; break;
            case TEL_WONT: t->t_State = TS_WONT; break;
            case TEL_DO:   t->t_State = TS_DO;   break;
            case TEL_DONT: t->t_State = TS_DONT; break;
            case TEL_SB:
                t->t_State  = TS_SB;
                t->t_SubLen = 0;
                break;
            default:
                /* NOP, DM, BRK, IP, AO, AYT, EC, EL, GA -- nothing useful for
                 * a BBS session, so drop them. */
                t->t_State = TS_DATA;
                break;
            }
            break;

        case TS_WILL: handle_will(t, c); t->t_State = TS_DATA; break;
        case TS_WONT: handle_wont(t, c); t->t_State = TS_DATA; break;
        case TS_DO:   handle_do(t, c);   t->t_State = TS_DATA; break;
        case TS_DONT: handle_dont(t, c); t->t_State = TS_DATA; break;

        case TS_SB:
            if (c == TEL_IAC)
                t->t_State = TS_SB_IAC;
            else if (t->t_SubLen < SUBNEG_SIZE)
                t->t_SubBuf[t->t_SubLen++] = c;
            break;

        case TS_SB_IAC:
            if (c == TEL_SE)
            {
                /* Subnegotiation complete.  We do not act on any of them, but
                 * consuming them correctly is what keeps the stream in sync. */
                t->t_State = TS_DATA;
            }
            else if (c == TEL_IAC)
            {
                if (t->t_SubLen < SUBNEG_SIZE)
                    t->t_SubBuf[t->t_SubLen++] = TEL_IAC;
                t->t_State = TS_SB;
            }
            else
            {
                t->t_State = TS_SB;
            }
            break;
        }
    }

    return o;
}

/*
 * Application bytes in, wire bytes out.
 *
 * Worst case expansion is 2x (every byte an IAC, or every byte a CR needing a
 * NUL), so the caller must supply outmax >= 2 * inlen.  All input is consumed.
 */
ULONG telnet_encode_t(struct Telnet *t, const UBYTE *in, ULONG inlen, UBYTE *out, ULONG outmax)
{
    ULONG          o = 0;
    ULONG          i;
    UBYTE          c;

    if (!t->t_Enabled)
    {
        ULONG len = (inlen < outmax) ? inlen : outmax;
        memcpy(out, in, len);
        return len;
    }

    for (i = 0; i < inlen; i++)
    {
        c = in[i];

        if (o + 2 > outmax)
            break;

        if (c == TEL_IAC)
        {
            out[o++] = TEL_IAC;
            out[o++] = TEL_IAC;
        }
        else if (!t->t_BinaryLocal && c == '\r')
        {
            /* NVT: a CR that is not part of CR LF must be followed by NUL. */
            out[o++] = '\r';
            if (i + 1 < inlen && in[i + 1] == '\n')
            {
                out[o++] = '\n';
                i++;
            }
            else
            {
                out[o++] = '\0';
            }
        }
        else
        {
            out[o++] = c;
        }
    }

    return o;
}

File: tests/host/telnet.py
#!/usr/bin/env python3
# SerialTCP -- a virtual serial device and modem daemon for AmigaOS.
# Copyright (C) 2026 Gordon Heydon
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.

"""
A small interactive telnet client, for calling a SerialTCP node from the host.

macOS has shipped without /usr/bin/telnet since High Sierra, and `nc` is no
substitute here: it does not answer option negotiation, so the session opens
with a spray of IAC bytes and stays in the wrong mode for ANSI or Zmodem.

This answers negotiation properly, agrees to 8-bit binary in both directions
(which is what a BBS needs), and puts the terminal in raw mode so keystrokes
reach the Amiga one at a time.

    tests/host/telnet.py [host] [port]        default 127.0.0.1 2323

Ctrl-]  quits.  Ctrl-C is passed through to the remote end, not to this client,
so it reaches the BBS the way it would over a real modem.
"""

import socket
import sys
import select
import termios
import tty

IAC, DONT, DO, WONT, WILL, SB, SE = 255, 254, 253, 252, 251, 250, 240
BINARY, ECHO, SGA = 0, 1, 3

QUIT_KEY = 0x1D          # Ctrl-]


class Telnet:
    def __init__(self, sock):
        self.s = sock
        self.buf = bytearray()
        self.state = "data"
        self.sub = bytearray()
        # Options we are willing to turn on. Everything else is refused.
        self.local = {BINARY: False, SGA: False}
        self.remote = {BINARY: False, SGA: False, ECHO: False}

    def send_cmd(self, verb, opt):
        self.s.sendall(bytes([IAC, verb, opt]))

    def offer(self):
        """Opening offer: ask for 8-bit clean both ways."""
        self.send_cmd(WILL, BINARY)
        self.send_cmd(DO, BINARY)
        self.send_cmd(DO, SGA)

    def _do(self, opt):
        if opt in self.local:
            if not self.local[opt]:
                self.local[opt] = True
                self.send_cmd(WILL, opt)
        else:
            self.send_cmd(WONT, opt)

    def _dont(self, opt):
        if self.local.get(opt):
            self.local[opt] = False
            self.send_cmd(WONT, opt)

    def _will(self, opt):
        if opt in self.remote:
            if not self.remote[opt]:
                self.remote[opt] = True
                self.send_cmd(DO, opt)
        else:
            self.send_cmd(DONT, opt)

    def _wont(self, opt):
        if self.remote.get(opt):
            self.remote[opt] = False
            self.send_cmd(DONT, opt)

    def feed(self, data):
        """Wire bytes in, application bytes out."""
        out = bytearray()

        for c in data:
            st = self.state

            if st == "data":
                if c == IAC:
                    self.state = "iac"
                else:
                    out.append(c)

            elif st == "iac":
                if c == IAC:
                    out.append(IAC)
                    self.state = "data"
                elif c in (DO, DONT, WILL, WONT):
                    self.state = {DO: "do", DONT: "dont", WILL: "will", WONT: "wont"}[c]
                elif c == SB:
                    self.sub.clear()
                    self.state = "sb"
                else:
                    self.state = "data"

            elif st in ("do", "dont", "will", "wont"):
                {"do": self._do, "dont": self._dont,
                 "will": self._will, "wont": self._wont}[st](c)
                self.state = "data"

            elif st == "sb":
                if c == IAC:
                    self.state = "sb_iac"
                else:
                    self.sub.append(c)

            elif st == "sb_iac":
                # Subnegotiations are consumed and ignored; consuming them
                # correctly is what keeps the stream in sync.
                self.state = "data" if c == SE else "sb"

        return bytes(out)

    def write(self, data):
        """Application bytes out, IAC doubled so 0xFF survives."""
        self.s.sendall(data.replace(b"\xff", b"\xff\xff"))


def main():
    host = sys.argv[1] if len(sys.argv) > 1 else "127.0.0.1"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 2323

    try:
        sock = socket.create_connection((host, port), timeout=10)
    except OSError as e:
        print(f"Could not connect to {host}:{port} -- {e}")
        return 1

    print(f"Connected to {host}:{port}.  Ctrl-] to quit.\r", flush=True)
    sock.settimeout(None)

    tn = Telnet(sock)
    tn.offer()

    stdin_fd = sys.stdin.fileno()
    isatty = sys.stdin.isatty()
    saved = termios.tcgetattr(stdin_fd) if isatty else None

    # When stdin is a pipe rather than a terminal it hits EOF immediately.
    # Quitting there would make the client useless in a script, so instead we
    # stop watching stdin and keep reading the remote until it closes or goes
    # quiet.
    stdin_open = True
    idle_limit = None if isatty else 5.0

    try:
        if isatty:
            tty.setraw(stdin_fd)

        while True:
            watch = [sock] + ([sys.stdin] if stdin_open else [])
            r, _, _ = select.select(watch, [], [], idle_limit)

            if not r:
                break               # idle timeout (scripted use only)

            if sock in r:
                data = sock.recv(4096)
                if not data:
                    break
                out = tn.feed(data)
                if out:
                    sys.stdout.buffer.write(out)
                    sys.stdout.buffer.flush()

            if stdin_open and sys.stdin in r:
                key = sys.stdin.buffer.raw.read(1) if isatty else sys.stdin.buffer.read(1)
                if not key:
                    stdin_open = False
                    continue
                if key[0] == QUIT_KEY:
                    break
                tn.write(key)

    finally:
        if isatty:
            termios.tcsetattr(stdin_fd, termios.TCSADRAIN, saved)
        sock.close()

    print("\r\nDisconnected.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
